File: Chapter-7-Operator-Overloading/004-a-sample-program-to-introduce-when-friend-function-cannot-be-used-to-overload-=-operator.cpp
#include <bits/stdc++.h>

using namespace std;

class coOrdinate
{
    int x, y;

public:
    // coOrdinate() {}
    // coOrdinate(int x, int y)
    // {
    //     this->x = x;
    //     this->y = y;
    // }
    // soluton in above comments

    coOrdinate operator=(coOrdinate ob2)
    {
        x = ob2.x;
        y = ob2.y;
        return *this;
    }
    void display()
    {
        cout << x << " " << y << endl;
    }
};

int main()
{
    coOrdinate ob1(10, 10), ob2;
    ob2 = ob1;
    ob2.display();
    return 0;
}

File: Chapter-7-Operator-Overloading/004-a-sample-program-to-introduce-when-friend-function-cannot-be-used-to-overload-=-operator.py
import copy

class CoOrdinate:
    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y

    def __copy__(self):
        return CoOrdinate(self.x, self.y)
    
    def display(self):
        print(self.x, self.y)

if __name__ == '__main__':
    ob1 = CoOrdinate(10, 20)
    ob2 = copy.copy(ob1) # shallow copy
    ob2.display() # 10 20

    ob3 = copy.deepcopy(ob1) # deep copy
    ob3.display() # 10 20